### PR TITLE
Add cachemask variant for fake_quantize_affine

### PR DESCRIPTION
**Pull request by andrewor14**
Summary: In QAT, we often wish to filter out the gradients corresponding to values outside the expected quantization range, for example:

```
# Forward
q = _quantize_affine_no_dtype_cast(...)
dq = _dequantize_affine_no_dtype_check(...)
mask = torch.logical_and((q >= quant_min), (q <= quant_max))

# Backward
grad = grad * mask
```

The existing `fake_quantize_affine` returns the dequantized values only, so callers do not have access to this mask. This commit adds the variant to this op that returns both the dequantized values and the mask, similar to
`fake_quantize_per_tensor_affine_cachemask` in core.

Test Plan:
python test/quantization/test_quant_primitives.py -k test_fake_quantize_affine_cachemask

### CODE REVIEW

File: torchao/quantization/quant_primitives.py [hunk 1]
```diff
@@ -411,6 +412,87 @@ def fake_quantize_affine(
...
+    returns an additional outlier mask for intermediate quantized values.
+
+    Returns:
+      A 2-tuple of (
+          final fake quantized values,
+          outlier mask for intermediate quantized values
+      )
+
+    Args:
+      input (torch.Tensor): original float32, float16 or bfloat16 Tensor
```
Comment (jerryzh168) [inline torchao/quantization/quant_primitives.py:453]: we can dedup the comments by: https://github.com/pytorch/pytorch/blob/06ebf87a1eca6c345f5e3e39b63c2ef487695043/torch/ao/quantization/observer.py#L582

e.g. :func:`~torchao.quantization.quant_primitives.fake_quantize_affine`?

File: torchao/quantization/quant_primitives.py [hunk 2]
```diff
@@ -411,6 +412,74 @@ def fake_quantize_affine(
...
+    quant_dtype: torch.dtype,
+    quant_min: Optional[int] = None,
+    quant_max: Optional[int] = None,
+    zero_point_domain: ZeroPointDomain = ZeroPointDomain.INT,
+) -> Tuple[torch.Tensor, torch.Tensor]:
+    """
+    General fake quantize op for quantization-aware training (QAT).
+    This is equivalent to calling `quantize_affine` + `dequantize_affine`
+    but without the dtype casts.
+
```
Comment (jerryzh168) [inline torchao/quantization/quant_primitives.py:442]: please add a section for `Args` and link to `fake_quantize_affine`

File: torchao/quantization/quant_primitives.py [hunk 3]
```diff
@@ -411,6 +412,74 @@ def fake_quantize_affine(
...
+    Note: Compared to `fake_quantize_affine`, this consumes more memory and
+    returns an additional outlier mask for intermediate quantized values.
+
+    Returns:
+      A 2-tuple of (
+          final fake quantized values,
+          outlier mask for intermediate quantized values
+      )
+
+    Please refer to :func:`~torchao.quantization.quant_primitives.fake_quantize_affine`
```
Comment (jerryzh168) [inline torchao/quantization/quant_primitives.py:452]: oh, I think we can move this to before `Returns`

**Review by jerryzh168 — APPROVED** [review]
had a comment on updating the docstring a bit

**Comment by helloyongyang** [comment]
Hello, I have a question regarding your PR.
We find that `fake_quantize_affine_cachemask` function returns a mask which is always `True`. Because there is a clamp op in `_quantize_affine_no_dtype_cast`. What is the purpose of the mask's presence here?

We also noticed that you wrote in the comments in `fake_quantize_affine_cachemask` function: "without the dtype casts." Could you explain what is meant by "dtype casts" here?
